jupyter-js-widgets/src/widget.ts

```diff
@@ -347,61 +342,71 @@ class WidgetModel extends Backbone.Model {
-                        this.msg_buffer = _.extend(this.msg_buffer || {}, attrs);
+                        this.msg_buffer = _.extend(this.msg_buffer || {}, msgState);
                         break;
                     case 'update':
                     case 'create':
-                        this.msg_buffer = attrs;
+                        this.msg_buffer = msgState;
                         break;
                     default:
-                        error();
-                        return false;
+                        throw 'unrecognized syncing method';
                 }
                 this.msg_buffer_callbacks = callbacks;
-
             } else {
                 // We haven't exceeded the throttle, send the message like
                 // normal.
                 this.send_sync_message(attrs, callbacks);
-                this.pending_msgs++;
                 // Since the comm is a one-way communication, assume the message
                 // arrived and was processed successfully.
                 // Don't call options.success since we don't have a model back from
                 // the server. Note that this means we don't have the Backbone
                 // 'sync' event.
-
             }
         }
     }
 
-    send_sync_message(attrs, callbacks) {
-        // prepare and send a comm message syncing attrs
-        // first, build a state dictionary with key=the attribute and the value
-        // being the value or the promise of the serialized value
-        var serializers = (this.constructor as typeof WidgetModel).serializers;
-        if (serializers) {
-            for (var k in attrs) {
-                if (serializers[k] && serializers[k].serialize) {
-                    attrs[k] = (serializers[k].serialize)(attrs[k], this);
-                }
+    /**
+     * Serialize widget state.
+     *
+     * A serializer is a function which takes in a state attribute and a widget,
+     * and synchronously returns a JSONable object. The returned object will
+     * have toJSON called if possible, and the final result should be a
+     * primitive object that is a snapshot of the widget state that may have
+     * binary array buffers.
+     */
+    serialize(state) {
+        const serializers = (this.constructor as typeof WidgetModel).serializers || {};
+        for (const k of state) {
+            if (serializers[k] && serializers[k].serialize) {
+                state[k] = (serializers[k].serialize)(state[k], this);
+            } else {
+                // the default serializer just deep-copies the object
+                // TODO: this won't work if the object is a primitive object with binary buffers!
+                // How should we handle those? One way is to declare a serializer for fields
+                // that could have binary that copies what fields it can, and makes a decision
+                // about whether to copy the ArrayBuffer
+                state[k] = JSON.parse(JSON.stringify(state[k]));
```

Comment:
Can we maybe have a warning here, say when an exception occurs, give a hint what may have gone wrong?